Updated Docs for LB6 (#72)
chickensoft-games/chickensoft_site: content/docs/logic_blocks/testing/testing_states.mdx

Each logic block state has an internal `context` which wraps the logic block it is attached to.

> Behind the scenes, calling the state's `Input`, `Output`, and `Get` methods will manipulate the context (and by extent the logic block). Attempting to call those state methods when not attached to a logic block results in the state throwing an exception. This most commonly happens if you reference one of those methods in the state's constructor instead of in the relevant lifecycle callbacks, since the state is not attached at the time it is constructed.

Reviewer: ```suggestion
Behind the scenes, calling the state's `Input`, `Output`, and `Get` methods require the state to be the *active* state. Attempting to call those state methods when the state is not the currently active state on a logic block results in the state throwing an exception. This most commonly happens if you reference one of those methods in the state's constructor instead of in the relevant lifecycle callbacks, since the state is not active in a logic block at the time it is constructed.
```

## 🥸 Using the State Tester

> LogicBlocks allows you to create a fake context for testing purposes. Call `state.Test()` to get a `StateTester` — a fake context with its own blackboard that lets you collect inputs and outputs produced by the state.

Reviewer: ```suggestion
When testing states, call `state.Test()` to get a `StateTester` — a utility with its own blackboard that lets you collect inputs and outputs produced by the state. State testers make it trivial to verify that a state behaves as expected.
```

```csharp
[Fact]
public void PlaysBeepingSoundOnEnter() {
  var state = new TimerState.PoweredOn.Beeping();

  // Create a state tester (fake context) for testing purposes.
  var tester = state.Test();

  // Simulate the state being entered.
  state.Enter();

  // Verify that the state produced the outputs we expect.
  tester.Outputs.ShouldBe([new Output.PlayBeepingSound()]);
}
```

You can also verify inputs added during testing in the same way via `tester.Inputs`.

Here's another example where we add a dependency to the tester's blackboard. This allows us to swap the dependencies a state uses during testing.

```csharp
[Fact]
public void ChangesDuration() {
  var state = new TimerState.PoweredOn.Idle();

  var tester = state.Test();

  // Put a value on the blackboard for the state to use.

>   tester.Set(new TimerLogic.Data() { Duration = 30.0d });

Reviewer: nested classes like Data need to be part of TimerState these days, as Data buckets are *also* intended to be scoped per state hierarchy now. too many references here to annotate by hand.

state.Exit();
```

> You can pass a `LogicBlockState` instance as the `previous` (for `Enter`) or `next` (for `Exit`) argument to control which callbacks in the type hierarchy are invoked. This is useful for testing a derived state in isolation without triggering base class callbacks.

Reviewer: I believe we removed this functionality -- no next/previous states should be given anywhere in the API.

Author: It's actually still there, I was using it when writing the new tests for GameDemo

```csharp
var state = new TimerState.PoweredOn.Idle();
var previousState = new TimerState.PoweredOn.Beeping();
var tester = state.Test();

// Passing a state that derives from PoweredOn as the "previous" state tells LogicBlocks
// that we're already in a PoweredOn context — so only the Idle-specific
// entrance callbacks (if any) will run, not the PoweredOn ones.
state.Enter(previousState);
```

<Callout type="info">

> Passing a previous/next state is optional and usually only done when testing a state that derives from another state you have separate unit tests for.

Reviewer: Again, removed this (or should).
